### PR TITLE
server: filter the EOF error for normal closed at handshake 

### LINKED ISSUE #31063: server: do not report `["encountered error"] [error=EOF]` for TCPConn alive detectors
## Enhancement

After https://github.com/pingcap/tidb/pull/27962, TiDB reports the handshake error if the connection is not successfully established, as a trivial bugfix. This change is released with v5.3.0

However, we found that some load balancers use the TCP connection for alive detectives, for example, the [GCP](https://cloud.google.com/load-balancing/docs/network) and https://asktug.com/t/topic/273508.

We can easily reproduce the following error logs from TiDB:

```
#!/usr/bin/env python    
    
import socket    
    
TCP_IP = '127.0.0.1'    
TCP_PORT = 4000    
BUFFER_SIZE = 1024    
MESSAGE = ""    
    
s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)    
s.connect((TCP_IP, TCP_PORT))    
s.close()    
```

You can see the following error logs from TiDB:
```
[ERROR] [terror.go:307] ["encountered error"] [error=EOF] [stack="github.com/pingcap/tidb/parser/terror.Log\n\t/home/bb7133/Projects/gopath/src/github.com/pingcap/tidb/parser/terror/terror.go:307\ngithub.com/pingcap/tidb/server.(*Server).onConn\n\t/home/bb7133/Projects/gopath/src/github.com/pingcap/tidb/server/server.go:519"]
```

The although harmless, the logs can be confusing to the users, so I think it's better to change the log level to 'DEBUG' when the `io.EOF` error is encountered: https://github.com/pingcap/tidb/blob/abb65829101d2f19b69b4e5a4dd0471dbed05915/server/server.go#L513

BTW, MySQL reports the following logs for the case above:
```
2021-12-27T14:03:16.793444Z 12 [Note] Got an error reading communication packets
```
So the failed connection logs happened for MySQL, but it is better in `Note` level.

### PR DESCRIPTION
**Pull request by knull-cn**
### What problem does this PR solve?

Issue Number: close #31063

Problem Summary: loader balance (eg: gcp lb )check health : construct tcp link ok and closed before handshake , tidb will print error log for EOF.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
  - run `tcp client` -- remember change the ip/port of tidb
    - `tcp client` test code was show latter
  - at tidb.log , you will see the `EOF` at debug level log and see other errors info in error level log
- [ ] No code

```
# `tcp client` test code
#!/usr/bin/env python    
import socket    
    
TCP_IP = '127.0.0.1'    
TCP_PORT = 4000    
BUFFER_SIZE = 1024    
MESSAGE = ""    
s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)    
s.connect((TCP_IP, TCP_PORT))    
s.close()
```

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by bb7133** [comment]
> Please follow PR Title Format:

> pkg [, pkg2, pkg3]: what is changed
> Or if the count of mainly changed packages are more than 3, use

> *: what is changed
> After you have format title, you can leave a comment /run-check_title to recheck it

@knull-cn Please fix the title check.

File: server/server.go [hunk 1]
```diff
@@ -513,7 +514,16 @@ func (s *Server) onConn(conn *clientConn) {
 		// Some keep alive services will send request to TiDB and disconnect immediately.
 		// So we only record metrics.
 		metrics.HandShakeErrorCounter.Inc()
-		terror.Log(errors.Trace(err))
+
+		if errors.Cause(err) == io.EOF {
+			// EOF is normal closed, it is not error
```
Comment (bb7133) [inline server/server.go:519]:
```suggestion
			// EOF is normally closed, we do not treat it as a noticable error and log it in 'DEBUG' level.
```
How about this?

**Comment by bb7133** [comment]
Please add a test case for your change, thanks!

File: server/server.go (hunk 1 above)
Comment (knull-cn) [reply server/server.go:519]: got it

File: server/server.go [hunk 2]
```diff
@@ -510,11 +511,17 @@ func (s *Server) onConn(conn *clientConn) {
...
-		// So we only record metrics.
-		metrics.HandShakeErrorCounter.Inc()
-		terror.Log(errors.Trace(err))
-		terror.Log(errors.Trace(conn.Close()))
+		if errors.Cause(err) == io.EOF {
+			// `EOF` means the connection is closed normally, we do not treat it as a noticeable error and log it in 'DEBUG' level.
+			logutil.BgLogger().With(zap.Uint64("conn", conn.connectionID)).
+				Debug("EOF", zap.String("remote addr", conn.bufReadConn.RemoteAddr().String()))
+		} else {
+			metrics.HandShakeErrorCounter.Inc()
```
Comment (knull-cn) [inline server/server.go:519]: counter the errors except `EOF`

Comment (jackysp) [reply server/server.go:519]: LGTM

File: server/server.go [hunk 3]
```diff
@@ -510,11 +511,17 @@ func (s *Server) onConn(conn *clientConn) {
...
+			logutil.BgLogger().With(zap.Uint64("conn", conn.connectionID)).
+				Error("Server.onConn handshake", zap.Error(err),
```
Comment (jackysp) [inline server/server.go:521]: I think `Warn` is enough, it is not an Error of tidb-server.

File: server/server.go [hunk 4]
```diff
@@ -510,11 +511,17 @@ func (s *Server) onConn(conn *clientConn) {
...
+					zap.String("remote addr", conn.bufReadConn.RemoteAddr().String()))
+		}
+		terror.Log(conn.Close())
```
Comment (jackysp) [inline server/server.go:524]: Why `errors.Trace` is removed.

Comment (knull-cn) [reply server/server.go:524]: Trace is `AddStack` to log info.  
There is not very depth , so i think file number of loggers is enough

Comment (jackysp) [reply server/server.go:524]: OK

**Review by djshow832 — APPROVED** [review]

**Comment by djshow832** [comment]
> Please add a test case for your change, thanks!

I think it's fine to only add a manual test because it only touches the logs.

**Review by jackysp — APPROVED** [review]
LGTM

**Pull request merged**